File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>&ans, vector<int>list, vector<int>& candidates, int index, int remain){
        // base condition
        if(remain<0) return;
        if(remain == 0){
            ans.push_back(list);
        }
        for(int i=index; i<candidates.size(); i++){
            list.push_back(candidates[i]);
            backtrack(ans, list, candidates, i, remain-candidates[i]);
            list.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>ans;
        vector<int>list;
        backtrack(ans,list,candidates,0,target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    private void helper(List<List<Integer>>ans, List<Integer>list, int index, int sum, int[] candidates, int target){
        if(sum == target){
            ans.add(new ArrayList(list));
            sum = 0;
            return;
        }
        if(sum>target){
            list = new ArrayList();
            sum = 0;
            return;
        }
        if(index == candidates.length) return;
        sum += candidates[index];
        list.add(candidates[index]);
        helper(ans, list, index, sum, candidates, target);
        list.remove(list.size()-1);
        sum -= candidates[index];
        helper(ans, list, index+1, sum, candidates, target);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>ans = new ArrayList();
        int n = candidates.length;
        helper(ans, new ArrayList(), 0, 0, candidates, target);
        return ans;
    }
}